File: Leetcode/minimum-add-to-make-parentheses-valid.py
class Solution:
    def minAddToMakeValid(self, S):
        """
        :type S: str
        :rtype: int
        """
        leftPar = 0
        rightPar = 0
        ans = 0
        for ch in S:
            if ch == '(':
                leftPar += 1
            elif ch == ")":
                if leftPar == 0:
                    ans += 1
                else: 
                    leftPar -= 1
        
        return ans + leftPar

File: Leetcode/minimum-add-to-make-parentheses-valid.cpp
class Solution {
public:
    int minAddToMakeValid(string S) {
        int len = S.size(), ans = 0;
        stack < char > st;
        for(int i = 0; i < len; i++){
            if (S[i] == '('){
                st.push(S[i]);
            }
            else{
                if(st.empty()){
                    ans++;
                }
                else{
                    st.pop();
                }
            }
        }
        while(!st.empty()){
            ans++;
            st.pop();
        }
        return ans;
    }
};